session/recursive.go

```diff
@@ -0,0 +1,105 @@
+package session
+
+import (
+	"sort"
+	"strings"
+
+	"launchpad.net/gozk"
+)
+
+var defaultACL = zookeeper.ACL{
+	Perms:  zookeeper.PERM_ALL,
+	Scheme: "world",
+	Id:     "anyone",
+}
+
+// ChildrenRecursive returns a slice all of a node's descendants that are at
+// most `maxDepth` levels away from the root.
+func (s *ZKSession) ChildrenRecursive(path string, maxDepth int) ([]string, error) {
+	rootDepth := 0
+	if path != "/" {
+		rootDepth = strings.Count(path, "/")
+	}
+
+	stat, err := s.Exists(path)
+	if err != nil {
+		return nil, err
+	}
+
+	if stat == nil {
+		return []string{}, nil
+	}
+
+	nodes := []string{path}
+	for index := 0; index < len(nodes); index++ {
+		if maxDepth <= 0 || strings.Count(nodes[index], "/") < rootDepth+maxDepth {
+			if children, _, err := s.Children(nodes[index]); err == nil {
+				parent := nodes[index]
+				if parent == "/" {
+					parent = ""
+				}
+
+				for _, child := range children {
+					nodes = append(nodes, parent+"/"+child)
+				}
+			}
+		}
+	}
+	return nodes[1:], nil
+}
+
+// SetRecursive will set data for the given path, creating all parents as
+// necessary with no data.
+func (s *ZKSession) SetRecursive(path string, data string) error {
+	// Since the Set method requires us to create intermediate nodes, we have to
+	// do a little extra work here
+	index := 0
+	for {
+		nextIndex := strings.Index(path[index+1:], "/")
+		if nextIndex < 0 {
+			break
+		}
+
+		index += nextIndex + 1
```

Comment:
`index += nextIndex` surprises me every time I see it. Maybe a comment here along the lines of `// nextIndex is an index into a suffix of the path`, or extract `path[index+1:]` as `pathSuffix` or something.